devtool-ide fixes

deribaucourt opened the pull request.

deribaucourt committed:
devtool: ide: add install dependency before gdbserver task

On VSCode, when running the debugger through launch.json, the program is
recompiled, gdbserver is started, and connected to. However, the task to
install the program on the target is not run, so the program is not
updated on the target.

Adding a dependency to the `install && deploy-target` task fixes this.

deribaucourt committed:
devtool: ide: remove editorconfig extension recommendation

Most packages do not provide a .editorconfig file. In those cases, the
recommended VSCode extension EditorConfig has no effect. Since devtool
ide is not producing a .editorconfig, it is not within our scope to
recommend this extension.

deribaucourt committed:
devtool: ide: recommend vscode cpp extension for meson

Just like CMake, Meson is usually used for C++ projects. We also
generate a c_cpp_properties.json file for Meson projects, so we should
recommend the C++ extension for VSCode.

deribaucourt committed:
devtool: ide: enable CMake configureOnOpen

In VSCode, the CMake extension must be configured for the intelliSense
engine to work. The user could already manually start this configuration
through the command palette.

afreof reviewed (commented):
nothing

afreof commented on scripts/lib/devtool/ide.py:
```diff
@@ -448,6 +450,10 @@ def initialize(self, config, workspace, tinfoil):
 
         self.dot_code_dir = os.path.join(self.srctree, '.vscode')
 
+        self.dot_editorconfig = os.path.join(self.srctree, '.editorconfig')
```
The .editorconfig files could be in any folder, not only in the top level. (See https://editorconfig.org/)

deribaucourt replied:
Usually these are placed at the root of the workspace and I'd like not to search the whole repository which could be slow. A compromise can be searching with only a small depth like 2. Yet, since this feature is really not critical, I find it reasonable to only look in the root. What do you think?

Otherwise, we could simply stop recommending this extension since it's not devtool ide which creates this .editorconfig. It's not a critical feature and although it's a good practice to have it, I don't think we should force it on users.

afreof replied:
Agree, not recommending it at all would be the best.

deribaucourt replied:
I've changed the commit to simply remove this extension.

afreof commented on scripts/lib/devtool/ide.py:
```diff
@@ -655,8 +664,9 @@ def __vscode_settings_cmake(self, settings_dict):
-        settings_dict["cmake.configureOnOpen"] = False
+        settings_dict["cmake.configureOnOpen"] = True
         settings_dict["cmake.sourceDirectory"] = self.real_srctree
+        settings_dict["cmake.cmakePath"] = os.path.join(self.staging_bindir_native, "cmake")
```
For cmake presets, the path is read by VSCode in the preset, not by cmake. So for the preset, the path should not be set twice. I haven't tested it, but theoretically the devtool ide plugin could run from multiple SDKs, with each SDK providing its own cmake-native.  (devtool modify -n allows you to specify the source folder). Selecting the defaults would just work. But it would break with this line.

In the case of cmake-kit, it's a bit more complicated. I suspect that this commit doesn't work because the devtool ide plugin doesn't create a workspace folder at all in the case of cmake-kit. It creates an entry in the user's global cmake tools configuration. And in the user's global configuration, we can't change the cmake executable globally.

The long story is here: https://github.com/microsoft/vscode-cmake-tools/issues/1242 At one point in development, I had an implementation for this trick here: https://github.com/microsoft/vscode-cmake-tools/issues/1242#issuecomment-1160608932 It works. But I removed it again.

I don't think this commit makes much sense yet. But I agree, using the cmake from the host is not perfect.

afreof replied:
Maybe I should improve the documentation about the cmake-kit use case.

```sh
devtool ide none core-image-minimal

mkdir kit-test
echo "project(foo VERSION 0.1)" > kit-test/CMakeLists.txt 
vscode kit-test
```

Ctrl + Shift + P --> cmake Quick Start --> Select kit

deribaucourt replied:
I agree to your conclusion. I did check that the preset was correctly using the bitbake CMake, however the kit doesn't. I think I'm not going to focus on the SHARED_SYSROOT_RECIPES for the moment since they are a out of the usual yocto workspace.

We could in the future generate a `.vscode` for the SHARED_SYSROOT_RECIPES in the future. We can already see the limitations of a CMake kit, and this will also be needed for other classes like meson.

afreof commented on scripts/lib/devtool/ide.py:
```diff
@@ -1030,6 +1049,9 @@ def vscode_tasks(self, args, gdb_cross):
                     "type": "shell",
                     "isBackground": True,
+                    "dependsOn": [
+                        "install && deploy-target %s" % self.bpn
```
Is it possible to press "F5" without initiating a deployment? I remember having that in the past as well.... I think I removed that dependency because the use case of "deploy once and run multiple debugging sessions" was more important to me than automatic deployment for the first debugging session.

But I agree that for users who haven't just implemented the devtool ide plugin themselves, it's significantly more intuitive with this dependency. Users who don't like the re-deployment can comment out the dependency. So I tend to change it according to your suggestion.

deribaucourt replied:
Yes that is closer to the traditional VSCode experience where changes in the code should be always reflected in the debugging session (the extension for pytest or nodejs do it for instance).

With my change however it is not possible to debug without deploying. But that would be risky since the executed code would not correspond to the current code if it was modified. Lines would not match for example. Please tell me if this breaks a specific use case for your original decision that we could consider supporting in other ways.

afreof replied:
The ideal solution would be to always call the deploy script (what your commit does) but also improve the generated deploy script to skip the deployment if nothing changed since the last deployment.

One idea would be to make the script which is generated by the function `gen_fakeroot_install_script` removing a stamp file and the Python script generated by "gen_deploy_target_script" writing a stamp file and only deploy if the stamp file does not exist.

This would not only speed up the workflow significantly. It would also allow that between two debugging sessions e.g. a config file on the device is changed without it being automatically overwritten again.

afreof replied:
maybe the  ``install && deploy-target %s" % self.bpn`` and deploy should be a variable. The task name and the dependency name must be equal.

deribaucourt replied:
> The ideal solution would be to always call the deploy script (what your commit does) but also improve the generated deploy script to skip the deployment if nothing changed since the last deployment.
> 
> One idea would be to make the script which is generated by the function `gen_fakeroot_install_script` removing a stamp file and the Python script generated by "gen_deploy_target_script" writing a stamp file and only deploy if the stamp file does not exist.
> 
> This would not only speed up the workflow significantly. It would also allow that between two debugging sessions e.g. a config file on the device is changed without it being automatically overwritten again.

it could speed things up for recipes producing huge binaries which are slow to install. I think this is rather a "nice to have" at this point.

deribaucourt replied:
> install && deploy-target %s" % self.bpn

Yes, it's a slight code duplication. I've updated the commit with a variable.

deribaucourt committed:
devtool: ide: check that gdbserver is installed

The VSCode debugging configuration launches gdbserver on the target.
Warn the user if gdbserver is not installed in the image.
This reproduces the behavior without IMAGE_GEN_DEBUGFS.

deribaucourt committed:
docs: sdk: list devtool ide dependencies

`devtool ide` requires specific packages and image configurations to be
installed. Document them in the SDK manual. devtool will also produce
warnings if the user tries to run `devtool ide` without the necessary
dependencies.

deribaucourt committed:
selftest: cpp-example: return 1 on failure

Running the test with ctest simply checks the return value of the
program.

afreof commented on documentation/sdk-manual/extensible.rst:
```diff
@@ -677,6 +677,23 @@ For a recipe which inherits cmake it does:
+   EXTRA_IMAGE_FEATURES += "debug-tweaks"
+
+   IMAGE_GEN_DEBUGFS = "1"
+   IMAGE_FSTYPES_DEBUGFS = "tar.bz2"
```
IMAGE_FSTYPES_DEBUGFS is not needed. The debug symbols are taken from the $WORKDIR of the image. Generating this bz2 can be a long running task with no value. (bz2 is even the slowest choice)

deribaucourt replied:
You are right! I tested and symbols are loaded without it. I had copied it from the doc around IMAGE_GEN_DEBUGFS. I removed this line.

afreof commented on scripts/lib/devtool/ide.py:
```diff
@@ -190,6 +191,8 @@ def initialize(self, config, tinfoil):
         self.combine_dbg_image = bb.data.inherits_class(
             'image-combined-dbg', image_d)
 
+        self.install_gdbserver = 'gdbserver' not in image_d.getVar('IMAGE_INSTALL')
```
This variable name is not good. What about ``gdbserver_missing``?

deribaucourt replied:
How did I not notice this! thanks

deribaucourt committed:
selftest: devtool: update ide cmake test strings

It seems that CMake logs with Yocto nambield are different. oe-selftest
ensure that the commands are run by looking for specific words in the
CMake output. I updated the test strings to match the new ones.

deribaucourt committed:
devtool: ide: update CmakeUserPresets.json

The file was overwritten unlike .vscode/*.json files which were merged
with what the original recipe provided, or user modifications.
However, the python dict.update() function does not merge nested
dictionaries, so some configurations will still be lost.

deribaucourt committed:
selftest: devtool: fix IMAGE_INSTALL for devtool ide tests

The IMAGE_INSTALL variable is now overriden in the image recipe
(oe-selftest-image.bb). We now use core-image-minimal as the base
like is done in othe SSH selftests like gcc.py and glibc.py.

deribaucourt committed:
devtool: ide: remove unused vscode_intelli_sense_mode()

This function is never called. The IntelliSense mode is configured
through configProvider in the c_cpp_properties.json file for CMake and
Meson classes.

This code could be usefull for future classes, but it is dead code for
now.

deribaucourt committed:
selftest: devtool: improve SSH and qemu arguments

selftest/gcc.py and glibc.py provide great examples on how to make
qemu testing more reliable, notably on headless machines. This patch
applies the same logic to the devtool tests.

deribaucourt committed:
devtool: ide: remove oe-local-files from vscode excludes

With the cmake-example recipe, the oe-local-files directory is
actually used when browsing or debugging the source code. The reason
is that VSCode automatically dereferences the symlinks present in the
workspace. Since the user will often be opening the symlinks and the
hard files, we should let him freely navigate them.

afreof commented on meta/lib/oeqa/selftest/cases/devtool.py:
```diff
@@ -2328,12 +2328,11 @@ def test_devtool_ide_recipe_cmake(self):
 
             # Verify re-building and testing works again
             result = runCmd('%s --build --preset %s --target clean' % (cmake_exe, preset_name), cwd=tempdir)
-            self.assertIn("Cleaning all built files...", result.output)
```
Not sure what's the issue. I just rebased my patches to the latest master and the tests pass without this patch here.

deribaucourt replied:
On my end I confirm that the cmake commands use a different output on my seutp.I'm getting:
```
2023-10-09 10:16:57,130 - oe-selftest - INFO - Traceback (most recent call last):
  File "/home/deribaucourt/Workspace/yocto-vscode/yocto/yocto-build/sources/poky/meta/lib/oeqa/selftest/cases/devtool.py", line 2337, in test_devtool_ide_recipe_cmake
    self.assertIn("Cleaning all built files...", result.output)
AssertionError: 'Cleaning all built files...' not found in "Change Dir: '/home/deribaucourt/Workspace/yocto-vscode/yocto/yocto-build/build-st/tmp/work/core2-64-poky-linux/cmake-example/1.0/cmake-example-1.0'\n\nRun Build Command(s): ninja -v clean\n[1/1] ninja  -t clean \nCleaning... 8 files."
```
This part especially looks different:
```
Run Build Command(s): ninja -v clean
[1/1] ninja  -t clean
Cleaning... 8 files.
```
My workspace has plain poky nanbield, with
 - cmake: 3.27.5
 - ninja: 1.11.1

afreof replied:
The tests are changed more or less as you suggested. The patches are rebased to todays master and they pass.

afreof commented on meta/lib/oeqa/selftest/cases/devtool.py:
```diff
@@ -2212,7 +2212,7 @@ def __devtool_ide_recipe(self, recipe_name, build_file, testimage):
             'IMAGE_CLASSES += "image-combined-dbg"',
             'IMAGE_GEN_DEBUGFS = "1"',
-            'IMAGE_INSTALL += "gdbserver %s-ptest"' % recipe_name
+            'PACKAGE_INSTALL:append = " gdbserver %s-ptest"\n' % recipe_name
```
not sure what't the issue here:

- Using the += operator as mentioned in the commit messages was fixed long time ago. The commit message looks outdated.
- Using PACKAGE_INSTALL instead of IMAGE_INSTALL add gdbserver also to special images like initramfs or the test-empty-image. This is probably not good.

I guess this commit is not needed.

deribaucourt replied:
The selftests were not passing on my nambield workspace because `oe-selftest-image.bb` overwrites IMAGE_INSTALL:
`IMAGE_INSTALL = "packagegroup-core-boot packagegroup-core-ssh-dropbear libudev"`. Hence, the original `IMAGE_INSTALL+=` statement was overshadowed and the packages not present on target. I don't understand how the test could work on your end however. Maybe you had modifications to the selftest image or core images?
I agree that PACKAGE_INSTALL is less pedantic. The later patchset uses core-image-minimal to stick to what's done in other runqemu SSH selftests.

afreof replied:
I looks like you are running a pretty old version of my patches. This has been fixed several weeks ago.

- [v6](https://patchwork.yoctoproject.org/project/oe-core/patch/20230910161850.4032227-12-adrian.freihofer@siemens.com/) is with append. It's from one month ago.
- [v5](https://patchwork.yoctoproject.org/project/oe-core/patch/20230905165913.2560907-14-adrian.freihofer@siemens.com/) has the bug

afreof replied:
Just to be sue, here is the latest from today Oct 9 2023: https://git.yoctoproject.org/poky-contrib/log/?h=adrianf/devtool-ide

afreof commented on documentation/sdk-manual/extensible.rst:
```diff
@@ -677,6 +677,22 @@ For a recipe which inherits cmake it does:
+   EXTRA_IMAGE_FEATURES += "ssh-server-openssh"
+   EXTRA_IMAGE_FEATURES += "debug-tweaks"
+
+   IMAGE_GEN_DEBUGFS = "1"
```
My setup looks like that:

```
# Build the companion debug file system
IMAGE_GEN_DEBUGFS = "1"
# But with devtool ide the dbg tar is not needed
IMAGE_FSTYPES_DEBUGFS = ""

# ssh is mandatory, no password simplifies the usage
EXTRA_IMAGE_FEATURES += "\
    ssh-server-openssh \
    debug-tweaks \
"

# Remote debugging needs the gdbserver on the target device
IMAGE_INSTALL:append = " gdbserver"

# Remote debugging works much better if the binaries are in the rootfs as well.
IMAGE_CLASSES += "image-combined-dbg"
```

deribaucourt replied:
Nice, I added the useful comments as well and removed the tar DEBUGFS.
Could you elaborate on what's not working well without image-combined-dbg? The debug symbols are not found when cross debugging?

afreof replied:
Yes if I remember correctly, gdb needs the binaries as well to find some debug symbols.

I have now revised the doc patch. I hope now everything is included and it is better understandable. Also the shared sysroot and cmake-kit part should now be covered.

afreof commented on scripts/lib/devtool/ide.py:
```diff
@@ -1074,6 +1077,10 @@ def setup_ide(self, args, image, gdb_cross):
 
+        if(image.gdbserver_missing):
+            logger.warning(
+                "gdbserver not installed in image. Remote debugging will not be available")
```
So far this plugin is used for formatting : https://github.com/microsoft/vscode-autopep8. This patch introduces 2 little issues.

deribaucourt replied:
Thanks for the hint, I ran the extension and fixed formatting.

afreof commented:
I pushed an update to my branch here: https://git.yoctoproject.org/poky-contrib/log/?h=adrianf/devtool-ide

Changes:

- Some of your findings I just fixed in my code (the trivial stuff)
  - remove editorconfig in my code
  - fix recommend vscode cpp extension for meson
- 6 Patches are cherry-picked
- Added 5 new commits
  - Added a fallback mode for recipes where we do not have native support like for cmake or meson
  - Make the arg parser smarter to handle e.g. something like that in one go: devtool ide cmake-example core-image-minimale meson-example
  - autotools support: needs much more testing
- Docs: restructured and improved the shared sysroots mode
- Tests: added a test for the shared sysroots mode

I think we could send a new version of the patchset including your patches. The last 5 from me are not ready yet.

deribaucourt commented:
> including your patches. The last 5 from me are not ready yet.

Yes, that sounds like a great plan! Thanks again for reviewing my patches and updating them to your newest branch.

I had a quick glance at your 5 WIP commits, they look great!

- I was wondering if we shouldn't add an abstraction for IDEs and recipe classes. As we add more classes like autotools, the ide.py file will start to get quite big, we could refactor it at some point into smaller units of code, each dedicated to an IDE or a bbclass while inheriting a common interface.
- For the autotools class, I don't know if the configurationProvider passes compiler flags (like defines, additional include paths...). Make sure to test if that works. If you already did, you can ignore my message...

afreof commented:
> * I was wondering if we shouldn't add an abstraction for IDEs and recipe classes.

Yes we should! It's not at the top at my proiority list but it is something which is needed.

> *  If you already did, you can ignore my message...

No I did not test it yet. I wrote the generator and run it. Then I improved my first blind try of the generator based on hints from the command line completion of VSCode. So yes, that would be helpful.

> * Yes, that sounds like a great plan!

Maybe a more concrete proposal for a plan. My assumption is that it will not be merged to poky master before end of month because poky is in feature freeze until then. So we should take the time to fix some details and bugs until then. My priority list:

- Imporve the parser (patch is ready for testing, changes something from a user's perspective)
- Fallback to devtool build and devtool deploy-target for recipes which are not natively support. (patch ready for testing, covers so many not handled cases)
- Handle "-native" recipes. I did not test that at all so far, but it's obvious that e.g. the remote debug configuration does not make any sense in the native context. Ideally we add check which terminates devtool ide with a reasonable error message for such nativesdk- or -native of recipes.
- Support multiple architectures. Use case: Change MACHINE variable and rebuild the IDE configuration would currently override the configuration for the old MACHINE. But that's not ideal. It would be better to add e.g. the PACKAGE_ARCH to the identifiers of the generated tasks and launch configurations.
- Fix what your patch "devtool: ide: update CmakeUserPresets.json" tries to fix, but recursively

Later on:
- Refactoring for modules IDE and modular Recipe support. This would also allow to e.g. add support for the clangd plugin to meta-clang or support for QTCreator to meta-qt
- Support more languages (Python, Rust, Go)

Just let me know in case you would like to take one of the items or feel free to start with a refactoring for more modular design.

deribaucourt commented:
> > * I was wondering if we shouldn't add an abstraction for IDEs and recipe classes.
> 
> Yes we should! It's not at the top at my proiority list but it is something which is needed.
> 
> > * If you already did, you can ignore my message...
> 
> No I did not test it yet. I wrote the generator and run it. Then I improved my first blind try of the generator based on hints from the command line completion of VSCode. So yes, that would be helpful.
> 
> > * Yes, that sounds like a great plan!
> 
> Maybe a more concrete proposal for a plan. My assumption is that it will not be merged to poky master before end of month because poky is in feature freeze until then. So we should take the time to fix some details and bugs until then. My priority list:
> 
>     * Imporve the parser (patch is ready for testing, changes something from a user's perspective)
> 
>     * Fallback to devtool build and devtool deploy-target for recipes which are not natively support. (patch ready for testing, covers so many not handled cases)
> 
>     * Handle "-native" recipes. I did not test that at all so far, but it's obvious that e.g. the remote debug configuration does not make any sense in the native context. Ideally we add check which terminates devtool ide with a reasonable error message for such nativesdk- or -native of recipes.
> 
>     * Support multiple architectures. Use case: Change MACHINE variable and rebuild the IDE configuration would currently override the configuration for the old MACHINE. But that's not ideal. It would be better to add e.g. the PACKAGE_ARCH to the identifiers of the generated tasks and launch configurations.
> 
>     * Fix what your patch "devtool: ide: update CmakeUserPresets.json" tries to fix, but recursively
> 
> 
> Later on:
> 
>     * Refactoring for modules IDE and modular Recipe support. This would also allow to e.g. add support for the clangd plugin to meta-clang or support for QTCreator to meta-qt
> 
>     * Support more languages (Python, Rust, Go)
> 
> 
> Just let me know in case you would like to take one of the items or feel free to start with a refactoring for more modular design.

Those features sound awesome! At the moment, I'm rather focused on the vscode-bitbake extension to call bitbake/devtool through the UI instead of terminals so I don't have the bandwidth to help with refactoring. Have you had feedback from the maintainers on merging the current patches? From my understanding, it would help them reviewing the patch list if we stick to the current feature set and make sure the code is easy to read. The CMake integration has now been thoroughly tested and improved by the both of us so I think it would be nice to first merge that. The other features you mention could get added in a second iteration.

The code is already nicely structured into functions, but if I could suggest anything, it would be to move around the code specific to VSCode and bbclasses in dedicated .py files/classes and have it separate from the common devtool ide logic. From the maintainers' eyes, it would mean it's going to be easier for the community to extend support to other IDEs.

BTW I had previously mentioned that the sysroot flags were not properly used in the cmake-example devtool ide workspace. This was actually a bug from the cmake extension which I've fixed in https://github.com/microsoft/vscode-cmake-tools/pull/3363

afreof commented:
> Have you had feedback from the maintainers on merging the current patches?

The tests passed on the Yocto AB and the patches are queued on the next branches from bootlin. Some simple patches are already on master.

> From my understanding, it would help them reviewing the patch list if we stick to the current feature set

Yes, that's my goal. I will send it as a new version of the same series. There will be a changelog in comparison to v6.

> if I could suggest anything, it would be to move around the code specific to VSCode and bbclasses in dedicated .py files/classes and have it separate from the common devtool ide logic.

Yes, there is room for improvement, I agree. But it's an "internal detail" which can be changed at any time without changing something from a user's perspective.

First priority is to make the user interface and the generated script namings more future proof. Without that the user has to delete the workspace and run devtool ide from scratch otherwise some old json files or scripts will not be properly replaced. If we would change the command line interface later on, also your VSCode plugin would no longer be compatible. We should really avoid that as much as we can.

> 
> BTW I had previously mentioned that the sysroot flags were not properly used in the cmake-example devtool ide workspace. This was actually a bug from the cmake extension which I've fixed in [microsoft/vscode-cmake-tools#3363](https://github.com/microsoft/vscode-cmake-tools/pull/3363)

Updated my setup to the version with your fix in the changelog. It works! Thank you.
